### PR TITLE
fix hmr for modules not referencing module.hot

### PR DESCRIPTION
**Pull request by JoviDeCroock**
Fixes: https://github.com/preactjs/wmr/issues/182

### CODE REVIEW

File: src/plugins/wmr/plugin.js [hunk 1]
```diff
@@ -75,15 +75,16 @@ export default function wmrPlugin({ hot = true } = {}) {
...
 			const s = new MagicString(code, {
 				filename: id,
 				// Typings from MagicString are wrong, see: https://github.com/Rich-Harris/magic-string/pull/182
 				// @ts-ignore
 				indentExclusionRanges: undefined
 			});
-			if (hot) {
+
+			if (!hasHot) {
+				s.prepend(`import { createHotContext as $w_h$ } from 'wmr'; $w_h$(import.meta.url);`);
```
Comment (developit) [inline src/plugins/wmr/plugin.js:86]:
```suggestion
				s.append(`\nimport { createHotContext as $w_h$ } from 'wmr'; $w_h$(import.meta.url);`);
```

File: src/plugins/wmr/plugin.js [hunk 2]
```diff
@@ -75,15 +77,16 @@ export default function wmrPlugin({ hot = true } = {}) {
...
 				// Typings from MagicString are wrong, see: https://github.com/Rich-Harris/magic-string/pull/182
 				// @ts-ignore
 				indentExclusionRanges: undefined
 			});
-			if (hot) {
+
+			if (!hasHot && !isProduction) {
+				s.append(`\nimport { createHotContext as $w_h$ } from 'wmr'; $w_h$(import.meta.url);`);
+			} else if (hot) {
 				s.append(after);
```
Comment (developit) [inline src/plugins/wmr/plugin.js:90]:
```suggestion
			if (hot) {
				if (!hasHot) {
					s.append(`\nimport { createHotContext as $w_h$ } from 'wmr'; $w_h$(import.meta.url);`);
				} else {
					s.append(after);
```

File: package.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
 	"bin": "wmr.cjs",
 	"type": "module",
 	"scripts": {
-		"demo": "cd demo && node --experimental-modules ../src/cli.js 2>&1",
+		"demo": "cd demo && node --experimental-modules ../src/cli.js --no-optimize --no-compress 2>&1",
```
Comment (developit) [inline package.json]: can we keep these? they just control gzip and node_modules optimization

File: src/plugins/wmr/plugin.js [hunk 4]
```diff
@@ -65,33 +66,41 @@ export default function wmrPlugin({ hot = true } = {}) {
 				before += `const module={${hot ? 'hot:import.meta.hot' : ''}};\n`;
 			}
 
+			const hasExport = code.match(/\bexport\b/);
```
Comment (developit) [inline src/plugins/wmr/plugin.js:69]: We can reuse the ESM_KEYWORDS regex from [here](https://github.com/preactjs/wmr/blob/7ede32763b18cd56b29a10cf79e884469c3a052d/src/plugins/fast-cjs-plugin.js#L3):

```suggestion
			const hasExport = code.match(/(\bimport\s*(\{|\s['"\w_$])|[\s;]export(\s+(default|const|var|let)[^\w$]|\s*\{))/);
```

File: src/plugins/wmr/plugin.js [hunk 5]
```diff
@@ -65,33 +66,41 @@ export default function wmrPlugin({ hot = true } = {}) {
...
-			if (code.match(/html`[^`]*<([a-zA-Z][a-zA-Z0-9.:-]*|\$\{.+?\})[^>]*>/) && code.match(/\bexport\b/)) {
+			if (code.match(/html`[^`]*<([a-zA-Z][a-zA-Z0-9.:-]*|\$\{.+?\})[^>]*>/) && hasExport) {
 				// if (this.getModuleInfo(id).hasJSX) {
 				hasHot = true;
 				after += '\n' + PREFRESH;
 				// }
 			}
 
-			if (!hasHot) return null;
+			if ((!hot && !hasHot) || (!hasExport && !code.match(/\bimport\b/))) return null;
```
Comment (developit) [inline src/plugins/wmr/plugin.js:81]: `hasHot` can be true in prod, but we will already have injected `null` for the HMR API on line 52, so we should still bail out here.

```suggestion
			if (!hot || !hasExport) return null;
```

File: src/plugins/wmr/plugin.js [hunk 6]
```diff
@@ -65,33 +66,41 @@ export default function wmrPlugin({ hot = true } = {}) {
...
-				s.append(after);
-				s.prepend(
-					`import { createHotContext as $createHotContext$ } from 'wmr';const $IMPORT_META_HOT$ = $createHotContext$(import.meta.url);${before}`
-				);
+				if (!hasHot) {
+					s.append(`\nimport { createHotContext as $w_h$ } from 'wmr'; $w_h$(import.meta.url);`);
+				} else {
+					s.append(after);
+					s.prepend(
+						`import { createHotContext as $createHotContext$ } from 'wmr';const $IMPORT_META_HOT$ = $createHotContext$(import.meta.url);${before}`
```
Comment (developit) [inline src/plugins/wmr/plugin.js:96]: might as well make these the same function as above:

```suggestion
						`import { createHotContext as $w_h $ } from 'wmr';const $IMPORT_META_HOT$ = $w_h $(import.meta.url);${before}`
```

File: package.json (hunk 3 above)
Comment (developit) [reply package.json:7]: it came back lol

File: src/plugins/wmr/plugin.js [hunk 7]
```diff
@@ -65,33 +67,42 @@ export default function wmrPlugin({ hot = true } = {}) {
...
-			if (code.match(/html`[^`]*<([a-zA-Z][a-zA-Z0-9.:-]*|\$\{.+?\})[^>]*>/) && code.match(/\bexport\b/)) {
+			if (code.match(/html`[^`]*<([a-zA-Z][a-zA-Z0-9.:-]*|\$\{.+?\})[^>]*>/) && hasExport) {
 				// if (this.getModuleInfo(id).hasJSX) {
 				hasHot = true;
 				after += '\n' + PREFRESH;
 				// }
 			}
 
-			if (!hasHot) return null;
+			if (!hot || !hasEsmKeywords) return null;
```
Comment (developit) [inline src/plugins/wmr/plugin.js:83]: third time's a charm?

```suggestion
			if (!hasHot && (!hot || !hasEsmKeywords)) return null;
```

**Review by marvinhagemeister — APPROVED** [review]

**Pull request merged**